ql/src/java/org/apache/hadoop/hive/ql/parse/repl/metric/MetricSink.java

```diff
@@ -0,0 +1,116 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.parse.repl.metric;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.api.ReplicationMetricList;
+import org.apache.hadoop.hive.metastore.api.ReplicationMetrics;
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf;
+import org.apache.hadoop.hive.metastore.utils.Retry;
+import org.apache.hadoop.hive.ql.metadata.Hive;
+import org.apache.hadoop.hive.ql.parse.repl.metric.event.ReplicationMetric;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * MetricSink.
+ * Scheduled thread to poll from Metric Collector and persists to DB
+ */
+public final class MetricSink {
+  private static final Logger LOG = LoggerFactory.getLogger(MetricSink.class);
+  private ScheduledExecutorService executorService;
+  private static volatile MetricSink instance;
+  private boolean isInitialised = false;
+  private HiveConf conf;
+
+  private MetricSink() {
+    this.executorService = Executors.newSingleThreadScheduledExecutor();
+  }
+
+  public static MetricSink getInstance() {
+    if (instance == null) {
+      synchronized (MetricSink.class) {
+        if (instance == null) {
+          instance = new MetricSink();
+        }
+      }
+    }
+    return instance;
+  }
+
+  public synchronized void init(HiveConf conf) {
+    if (!isInitialised) {
+      this.conf = conf;
+      long frequencyInMins = MetastoreConf.getLongVar(conf, MetastoreConf.ConfVars.REPL_METRICS_UPDATE_FREQUENCY);
+      this.executorService.schedule(new MetricSinkWriter(conf), frequencyInMins, TimeUnit.MINUTES);
+      isInitialised = true;
+    }
+  }
+
+  static class MetricSinkWriter implements Runnable {
+    private MetricCollector collector;
+    private HiveConf conf;
+
+    // writer instance
+
+    MetricSinkWriter(HiveConf conf) {
+      this.collector = MetricCollector.getInstance();
+      this.conf = conf;
+    }
+
+    @Override
+    public void run() {
+      // get metrics
+      List<ReplicationMetric> metrics = collector.getMetrics();
+      // write metrics and retry if fails
+      Retry<Void> retriable = new Retry<Void>(Exception.class) {
+        @Override
+        public Void execute() throws Exception {
+          ReplicationMetricList metricList = new ReplicationMetricList();
+          List<ReplicationMetrics> replicationMetricsList = new ArrayList<>();
```

Comment:
Can we initialize the new ArrayList<>() with same size as list 'metrics'. This will save on resizing cost which includes Array copy operation.